File: src/main/java/ir/sayandevelopment/listener/TabCompleteListener.java
package ir.sayandevelopment.listener;

import com.velocitypowered.api.event.Subscribe;
import com.velocitypowered.api.event.player.TabCompleteEvent;
import ir.sayandevelopment.VanishedPlayer;
import ir.sayandevelopment.command.VanishCommand;

import java.util.Collection;
import java.util.List;

public class TabCompleteListener {

    @Subscribe
    public void onTabComplete(TabCompleteEvent event) {
        List<String> suggestions = event.getSuggestions();
        Collection<VanishedPlayer> vanishedPlayers = VanishCommand.vanishedPlayers.values();
        for (VanishedPlayer vanishedPlayer : vanishedPlayers) {
            if (vanishedPlayer.isVanished()) {
                suggestions.removeIf(suggestion -> suggestion.equalsIgnoreCase(vanishedPlayer.getUserName()));
            }
        }
    }
}

File: src/main/java/ir/syrent/velocityvanish/velocity/listener/TabCompleteListener.kt
package ir.syrent.velocityvanish.velocity.listener

import com.velocitypowered.api.event.Subscribe
import com.velocitypowered.api.event.player.TabCompleteEvent
import ir.syrent.velocityvanish.velocity.VelocityVanish
import ir.syrent.velocityvanish.velocity.vruom.VRuom

class TabCompleteListener(
    private val plugin: VelocityVanish
) {

    init {
        VRuom.registerListener(this)
    }

    @Subscribe
    private fun onTabComplete(event: TabCompleteEvent) {
        if (event.player.hasPermission("velocityvanish.bypass.tabcompleter")) return
        
        event.suggestions.removeIf {
            plugin.vanishedPlayers.contains(it)
        }
    }
}